octopusdeploy/lifecycles.go

```diff
@@ -0,0 +1,204 @@
+		sling: sling,
+	}
+}
+
+type Lifecycles struct {
+	Items []Lifecycle `json:"Items"`
+	PagedResults
+}
+
+type Lifecycle struct {
+	ID                      string          `json:"Id,omitempty"`
+	Name                    string          `json:"Name" validate:"required"`
+	Description             string          `json:"Description,omitempty"`
+	ReleaseRetentionPolicy  RetentionPeriod `json:"ReleaseRetentionPolicy"`
+	TentacleRetentionPolicy RetentionPeriod `json:"TentacleRetentionPolicy"`
+	Phases                  []Phase         `json:"Phases"`
+}
+
+type RetentionUnit string
+
+const (
+	RetentionUnit_Days = "Days"
+	RetentionUnit_Items = "Items"
+)
+
+type RetentionPeriod struct {
+	Unit              RetentionUnit `json:"Unit"`
+	QuantityToKeep    int32         `json:"QuantityToKeep"`
+	ShouldKeepForever bool          `json:"ShouldKeepForever"`
+}
+
+type Phase struct {
+	ID                                 string          `json:"Id,omitempty"`
+	Name                               string          `json:"Name" validate:"required"`
+	MinimumEnvironmentsBeforePromotion int32           `json:"MinimumEnvironmentsBeforePromotion"`
+	IsOptionalPhase                    bool            `json:"IsOptionalPhase"`
+	ReleaseRetentionPolicy             *RetentionPeriod `json:"ReleaseRetentionPolicy"`
+	TentacleRetentionPolicy            *RetentionPeriod `json:"TentacleRetentionPolicy"`
+	AutomaticDeploymentTargets         []string        `json:"AutomaticDeploymentTargets"`
+	OptionalDeploymentTargets          []string        `json:"OptionalDeploymentTargets"`
+}
+
+
+func NewLifecycle(name string) *Lifecycle {
+	return &Lifecycle{
+		Name:   name,
+		Phases: []Phase{},
+		TentacleRetentionPolicy:RetentionPeriod{
+			Unit:RetentionUnit_Days,
+		},
+		ReleaseRetentionPolicy:RetentionPeriod{
+			Unit:RetentionUnit_Days,
+		},
+	}
+}
+
+// ValidateLifecycleValues checks the values of a Lifecycle object to see if they are suitable for
+// sending to Octopus Deploy. Used when adding or updating lifecycles.
+func ValidateLifecycleValues(Lifecycle *Lifecycle) error {
+	validate := validator.New()
+
+	err := validate.Struct(Lifecycle)
+
+	if err != nil {
+		return err
+	}
+
+	if Lifecycle.Phases != nil {
+		for _, phase := range Lifecycle.Phases {
+			phaseErr := validate.Struct(phase)
+
+			if phaseErr != nil {
+				return phaseErr
+			}
+		}
+	}
+
+	return nil
+}
+
+// Get returns a single lifecycle by its lifecycleid in Octopus Deploy
+func (s *LifecycleService) Get(LifecycleID string) (*Lifecycle, error) {
+	path := fmt.Sprintf("lifecycles/%s", LifecycleID)
+	resp, err := apiGet(s.sling, new(Lifecycle), path)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return resp.(*Lifecycle), nil
+}
+
+// GetAll returns all lifecycles in Octopus Deploy
+func (s *LifecycleService) GetAll() (*[]Lifecycle, error) {
+	return s.get("");
+}
+
+func (s *LifecycleService) get(query string) (*[]Lifecycle, error) {
+	var p []Lifecycle
+
+	path := "lifecycles?take=2147483647"
```

Comment:
Save Multiple requests if the server supports it

Reply:
I will accept this method for now but I want to standardize on this at some point: https://github.com/MattHodge/go-octopusdeploy/issues/34